### PR TITLE
Theme

### PR DESCRIPTION
**Pull request by Patchethium**
## 内容

Add theme. At current stage we have only two themes Default and Dark, but anyone can add one easily by creating a json file in `public/themes` referring to other theme config files or the template below. The color format, if not specified, is hex:
```json
{
  "name": "theme name",
  "isDark": "is it a dark theme",
  "colors": {
    "--color-primary": "the primary color",
    "--color-primary-rgb": "primary color in rgb",
    "--color-primary-light": "color for the sliders and scrollbar",
    "--color-primary-light-rgb": "primary-light in rgb",
    "--color-display": "text color",
    "--color-display-rgb": "text color in rgb",
    "--color-display-light": "light text color",
    "--color-display-dark": "dark text color",
    "--color-background": "background color",
    "--color-background-light": "light background, used in buttons on head bar",
    "--color-setting-item": "setting items color",
    "--color-warning": "warning color",
    "--color-markdown-color": "the text color of markdown pages",
    "--color-markdown-background": "the background color of markdown pages",
    "--color-markdown-hyperlink": "the hyperlink color of markdown pages"
  }
}
```

As you can see, I removed the `secondary` to `display` for the keyword is not used to showing texts. I'm also considering an interactive theme builder in the future.

## 関連 Issue

ref #316

## その他

I may need some help in themes of Markdown pages from who wrote it, I just don't feel like digging into that part. Just tell me what to do and I'll add it to the code. Before that I'll keep this pr a draft.

### CODE REVIEW

**Comment by takana-v** [comment]
github-markdown-cssが色を上書きしているようです。
このコードを追加するとうまくいくかもしれません。
It seems that github-markdown-css is overriding the colors.
Adding this code might work.

(_index.scss)
```css
.markdown-body {
  color: var(--color-display);
}
```

**Comment by Hiroshiba** [comment]
I think it's an interesting feature!

**Comment by Patchethium** [comment]
@takana-v That makes sense. But how can I dynamically change the external css file source in javascript? I made a few search but unfortunately none of the answers on stack overflow actually worked.

BTW, the background color from the github markdown css turns out to be only filling the space up to one page, any idea of fixing this?

image

**Comment by takana-v** [comment]
I think it is difficult to modify the github markdown css.
How about overriding the text and background colors by `_index.scss`?
```css
.markdown-body {
  color: var(--color-display);
  background: var(--color-background);
}
```
image

**Comment by Patchethium** [comment]
So this means we're getting rid of the github markdown and write the color ourselves, the template above is updated and the following colors are added:

```json
  "color-markdown-color": "the text color of markdown pages",
  "color-markdown-background": "the background color of markdown pages",
  "color-markdown-hyperlink": "the hyperlink color of markdown pages"
```

If that's not a problem it should be good to go now.

File: public/themes/dark.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+  "name": "Dark",
+  "isDark": true,
+  "colors": {
+    "--color-primary": "#0A0A0A",
```
Comment (Hiroshiba) [inline public/themes/dark.json:5]: VOICEVOXの（ダークモード版の）プライマリーカラーの適用に挑戦したいと思っています。
構いませんか？ :-)
I would like to try to apply the ( dark mode version of ) primary color of VOICEVOX.
May I? :-)

File: src/App.vue [hunk 2]
```diff
@@ -19,12 +19,19 @@ export default defineComponent({
...
+    promises.push(store.dispatch("GET_USE_GPU", undefined));
+    promises.push(store.dispatch("GET_INHERIT_AUDIOINFO"));
+    promises.push(store.dispatch("GET_SAVING_SETTING"));
+    promises.push(store.dispatch("GET_HOTKEY_SETTINGS"));
+    promises.push(store.dispatch("GET_USE_VOICING"));
+    promises.push(store.dispatch("GET_THEME_SETTING"));
+
+    Promise.all(promises).then(() => {
+      store.dispatch("VUEX_READY", undefined);
+    });
```
Comment (Hiroshiba) [inline src/App.vue:34]: 新しく`START_VUEX` actionを作って、これらの処理を全部`START_VUEX`内に書くのはどうでしょう？
How about making a new `START_VUEX` action and put all these processes in `START_VUEX`?

File: package-lock.json [hunk 3]
```diff
@@ -19681,6 +19624,87 @@
         }
       }
     },
+    "vue-loader-v16": {
+      "version": "npm:vue-loader@16.8.3",
+      "resolved": "https://registry.npmjs.org/vue-loader/-/vue-loader-16.8.3.tgz",
+      "integrity": "sha512-7vKN45IxsKxe5GcVCbc2qFU5aWzyiLrYJyUuMz4BQLKctCj/fmCa0w6fGiiQ2cLFetNcek1ppGJQDCup0c1hpA==",
+      "dev": true,
+      "optional": true,
+      "requires": {
```
Comment (Hiroshiba) [inline package-lock.json:19633]: packageから`github-markdown-css`が消えているのに、.lockに何かが追加されているのが気になりました。
I noticed that `github-markdown-css` was removed from the package, but something was added to the .lock.

**Review by Hiroshiba — CHANGES_REQUESTED** [review]
It's so great!!! Nice work!!

> I'm also considering an interactive theme builder in the future.

That's very cool.
What kind of things will you be making?

File: package-lock.json (hunk 3 above)
Comment (Patchethium) [reply package-lock.json:19633]: It's npm's work, shouldn't be a problem. If you're not sure about that, you can run `npm uninstall github-markdown-css` on your own machine and copy the package-lock.json here after merging.

File: src/store/setting.ts [hunk 4]
```diff
@@ -104,6 +120,41 @@ export const settingStore: VoiceVoxStoreOptions<
...
+    },
+    SET_THEME_SETTING(
+      { state, commit },
+      { currentTheme }: { currentTheme: string }
+    ) {
+      window.electron.theme(currentTheme);
+      const theme = state.themeSetting.availableThemes.find((value) => {
+        return value.name == currentTheme;
+      });
+      if (theme) {
```
[the remark this answers is not in the document]
Comment (Patchethium) [reply src/store/setting.ts:143]: I'm not sure, similar code is everywhere, I don't think it's necessary to use a 3rd party module for a type checking.

[the remark this answers is not in the document]
Comment (Hiroshiba) [reply src/store/setting.ts:143]: あ、すみません！assertではなく`if () raise Error`でした！
Oh, sorry, it was `if () raise Error`, not `assert`!

File: public/themes/dark.json (hunk 1 above)
Comment (Patchethium) [reply public/themes/dark.json:5]: Suit yourself, the reason I put themes in experimental function is I haven't decided the colors.

どうぞ、そもそも実験的な機能に置いたのも私が色を決められないためです。

**Comment by Patchethium** [comment]
> That's very cool.
> What kind of things will you be making?

Something just like [quasar's](https://quasar.dev/style/theme-builder#introduction) theme builder, maybe with a light weight preview and export/import function.

File: package-lock.json (hunk 3 above)
Comment (Hiroshiba) [reply package-lock.json:19633]: Okay!

File: src/type/preload.d.ts [hunk 5]
```diff
@@ -146,3 +148,49 @@ export type MoraDataType =
   | "pitch"
   | "pause"
   | "voicing";
+
+export type ThemeConf = {
+  name: string;
+  isDark: boolean;
+  colors: {
```
Comment (Hiroshiba) [inline src/type/preload.d.ts:155]: TypeScriptに詳しくないのですが、もしかしたらここは以下のようにすることで、１回の定義で済ませられるかもしれません。
I'm not familiar with TypeScript, but you might be able to change this here.
```ts
{ [K in ThemeColorType]: string }
```

**Review by Hiroshiba — APPROVED** [review]
LGTM！！

エクスポート・インポート機能！！素晴らしいと思います！
軽量なプレビュー機能もあるととても便利だと思います、良いですね！
Export/Import feature! I think it's great!
I think a lightweight preview feature would be very useful too, good!

またのプルリクエストをお待ちしております。
We look forward to your pull requests again.

**Pull request merged**